fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggThroughJoinOnPkFk.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    private @Nullable List<NamedExpression> constructNewOutput(LogicalAggregate<?> agg, Set<Slot> primaryOutput,
+            Map<Slot, Slot> primaryToForeignDeps, FuncDeps funcDeps, Plan primaryPlan) {
+        List<NamedExpression> newOutput = new ArrayList<>();
+        for (NamedExpression expression : agg.getOutputExpressions()) {
+            // There are three cases for output expressions:
+            // 1. Slot: the slot is from primary plan, we need to replace it with
+            //             the corresponding slot from foreign plan,
+            //             or skip it when it isn't in group by.
+            // 2. Count: the count is from primary plan,
+            //             we need to replace the slot in the count with the corresponding slot
+            //             from foreign plan
+            if (expression instanceof Slot && primaryPlan.getOutput().contains(expression)) {
+                if (primaryToForeignDeps.containsKey(expression)) {
+                    expression = primaryToForeignDeps.getOrDefault(expression, expression.toSlot());
+                } else {
+                    continue;
+                }
+            }
+            if (expression instanceof Alias
+                    && expression.child(0) instanceof Count
+                    && expression.child(0).child(0) instanceof Slot) {
+                // count(slot) can be rewritten by circle deps
+                Slot slot = (Slot) expression.child(0).child(0);
+                if (primaryToForeignDeps.containsKey(slot)
+                        && funcDeps.isCircleDeps(
+                                ImmutableSet.of(slot), ImmutableSet.of(primaryToForeignDeps.get(slot)))) {
+                    expression = (NamedExpression) expression.rewriteUp(e ->
+                            e instanceof Slot
+                                    ? primaryToForeignDeps.getOrDefault((Slot) e, (Slot) e)
+                                    : e);
+                }
+            }
+            if (!(expression instanceof Slot)
+                    && expression.getInputSlots().stream().anyMatch(primaryOutput::contains)) {
+                return null;
+            }
+            newOutput.add(expression);
+        }
+        return newOutput;
+    }
+
+    // try to extract primary key table and foreign key table
+    private @Nullable Pair<Plan, Plan> tryExtractPrimaryForeign(LogicalJoin<?, ?> join) {
+        Plan primary;
+        Plan foreign;
+        if (JoinUtils.canEliminateByFk(join, join.left(), join.right())) {
+            primary = join.left();
+            foreign = join.right();
+        } else if (JoinUtils.canEliminateByFk(join, join.right(), join.left())) {
+            primary = join.right();
+            foreign = join.left();
+        } else {
+            return null;
+        }
+        return Pair.of(primary, foreign);
+    }
+
+    /**
+     * This class flattens nested join clusters and optimizes aggregation pushdown.
```

Comment:
currently it is a customized support of rbo-level join reorder. In the future we may extend more common utils to support other scenarios, such as over-64 join reorder by heristic methods.